Fix memory leak while pushing to MVs w/o query context (from Kafka/...)

azat opened the pull request:
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix memory leak while pushing to MVs w/o query context (from Kafka/...)

Patchset description:
- Avoid leaking of ThreadGroupStatus::finished_threads_counters_memory
- WriteBufferFromS3: remove unused ThreadGroupStatus
- Avoid using of ThreadGroupStatus of the MainThreadStatus
- Fix possible missing detachQuery() in case of exception in readers
- Fix memory leak while pushing to MVs w/o query context (from Kafka/...)

*You will find more details in the commit descriptions*

This is the [pdf](https://github.com/ClickHouse/ClickHouse/files/9440552/profile2.focus.pdf) jemalloc report before this patch series, for the 500MVs case, with this patches the memory usage does not goes beyond ~750MiB.

Fixes: #36367 (cc @kitaisreal @den-crane @lichenglin @yakov-olkhovskiy )

azat committed:
Fix memory leak while pushing to MVs w/o query context (from Kafka/...)

While pushign to MVs, there is a low-level code that create
ThreadGroupStatus/ThreadStatus, it is required to gather some metrics
for system.query_views_log.

But, one should not use ThreadGroupStatus of the MainThreadStatus, since
this structure can hold some state, that may not be cleaned, plus this
may be racy, instead it is better to create new ThreadGroupStatus and
attach it instead.

Also this place misses detachQuery(), and because of this it leaks
ThreadGroupStatus::finished_threads_counters_memory. But it is only the
problem pushing to MVs is done w/o query context (i.e. from Kafka/...),
since when it has query context detachQuery() will be called eventually.

Before this patch series, when I've tried the reproducer with
500 MVs attached to Kafka engine (that @den-crane suggested), jemalloc
report looks like this:

    $ ../jeprof --text ~/ch/tmp/upstream/clickhouse-binary --base jeprof.44384.0.i0.heap jeprof.44384.167.i167.heap
    Using local file /home/azat/ch/tmp/upstream/clickhouse-binary.
    Using local file jeprof.44384.167.i167.heap.
    Total: 915.6 MB
       910.7  99.5%  99.5%    910.7  99.5% Snapshot (inline)
         9.5   1.0% 100.5%      9.5   1.0% std::__1::__libcpp_operator_new (inline)
         0.5   0.1% 100.6%      0.5   0.1% DB::TasksStatsCounters::create

And with focus to this place:

    $ ../jeprof --focus Snapshot --text ~/ch/tmp/upstream/clickhouse-binary --base jeprof.44384.0.i0.heap jeprof.44384.167.i167.heap
    Using local file /home/azat/ch/tmp/upstream/clickhouse-binary.
    Using local file jeprof.44384.167.i167.heap.
    Total: 915.6 MB
       910.7 100.0% 100.0%    910.7 100.0% Snapshot (inline)
         0.0   0.0% 100.0%    910.7 100.0% DB::QueryPipeline::reset
         0.0   0.0% 100.0%    910.7 100.0% DB::StorageKafka::streamToViews
         0.0   0.0% 100.0%    910.7 100.0% DB::StorageKafka::threadFunc
         0.0   0.0% 100.0%    910.7 100.0% ProfileEvents::Counters::getPartiallyAtomicSnapshot
         0.0   0.0% 100.0%    910.7 100.0% ~ThreadStatus
         0.0   0.0% 100.0%    910.7 100.0% ~ViewRuntimeData
         0.0   0.0% 100.0%    910.7 100.0% ~ViewRuntimeStats (inline)

Actually this report does not looks great (you understand it because I
stripped it), because --text does not that smart, but if you will use
--pdf for the report you will see the stacktrace (will attach pdf to the
pull request).

But after this patch series the process RSS does not goes beyond
~700MiB.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Avoid leaking of ThreadGroupStatus::finished_threads_counters_memory

Cleanup them in ThreadStatus::detachQuery(), anyway they cannot be
received after by the client.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
>skip mergeable check

To ignore style check, due to #40733

kitaisreal commented on src/Disks/IO/ThreadPoolReader.cpp:
```diff
@@ -200,9 +200,9 @@ std::future<IAsynchronousReader::Result> ThreadPoolReader::submit(Request reques
-            ? CurrentThread::get().getThreadGroup()
-            : MainThreadStatus::getInstance().getThreadGroup();
+    ThreadGroupStatusPtr running_group;
+    if (CurrentThread::isInitialized() && CurrentThread::get().getThreadGroup())
```
I am not familiar with this place. Why do not we use deleter below ?
```
if (query_context)
    thread_status.attachQueryContext(query_context);
if (running_group)
    thread_status.attachQuery(running_group);
```
If exception will be thrown somewhere between attach and detach will we leak something ?
```
if (running_group)
    thread_status.detachQuery();
```

azat replied:
>If exception will be thrown somewhere between attach and detach will we leak something ?

You are right.

>I am not familiar with this place. Why do not we use deleter below ?

I will use `SCOPE_EXIT` here instead.

*`deleter` is not a common thing to use in a code base, however it is required in `buildPushingToViewsChain()` because it should capture some pointers*

kitaisreal commented on src/Disks/IO/ThreadPoolRemoteFSReader.cpp:
```diff
@@ -40,9 +40,9 @@ ThreadPoolRemoteFSReader::ThreadPoolRemoteFSReader(size_t pool_size, size_t queu
-            ? CurrentThread::get().getThreadGroup()
-            : MainThreadStatus::getInstance().getThreadGroup();
+    ThreadGroupStatusPtr running_group;
+    if (CurrentThread::isInitialized() && CurrentThread::get().getThreadGroup())
```
Same question as above.

kitaisreal commented on src/Interpreters/ThreadStatusExt.cpp:
```diff
@@ -406,6 +406,11 @@ void ThreadStatus::detachQuery(bool exit_if_already_detached, bool thread_exits)
+
+    /// Avoid leaking of ThreadGroupStatus::finished_threads_counters_memory
+    /// (this is in case someone uses system thread but did not call getProfileEventsCountersAndMemoryForThreads())
+    thread_group->getProfileEventsCountersAndMemoryForThreads();
```
Need to rename `getProfileEventsCountersAndMemoryForThreads` in something like `getProfileEventsCountersAndMemoryForThreads`, because it is unexpected that get method will cleanup something.

azat replied:
Egh, seems that you made a typo here, I will rename it to `resetProfileEventsCountersAndMemoryForThreads`

P.S. Usually I'm trying to keep bug fixes patches as small as possible to avoid possible conflicts during backporting.

kitaisreal commented on src/Processors/Transforms/buildPushingToViewsChain.cpp:
```diff
@@ -238,29 +238,30 @@ Chain buildPushingToViewsChain(
+        /// N times more interruptions
+        view_thread_status_ptr->disableProfiling();
+        /// view_thread_status_ptr will be moved later (on and on), so need to capture raw pointer.
+        view_thread_status_ptr->deleter = [thread_status = view_thread_status_ptr.get(), running_group]
```
Great catch. Why do we not always detachQuery in destructor ? It seems like here, we catch raw pointer to ourselves and use detach manually.

azat replied:
>Great catch. Why do we not always detachQuery in destructor ? It seems like here, we catch raw pointer to ourselves and use detach manually.

Usually we do, see `CurrentThread::initializeQuery` and friends, however this place uses pretty low level code (for ClickHouse source base).

I will take a look later on all of this, and maybe will come up with some more safe and generic approach.
But for now it is better to keep this patch small, to make it easy to backport.

kitaisreal reviewed (changes requested):
Changes looks good. Need to clarify some questions.

kitaisreal reviewed (approved):
Good explanations. Lets not rename any methods, only fix leaks and backport it.
We can rename methods in separate pull request.

azat committed:
WriteBufferFromS3: remove unused ThreadGroupStatus

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Avoid using of ThreadGroupStatus of the MainThreadStatus

One should not use MainThreadStatus, since ThreadGroupStatus can hold
some states, and it is better not to play with this, since this may
create leaks.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix possible missing detachQuery() in case of exception in readers

This can create leaks, since detachQuery() responsible for cleaning,
i.e. ThreadGroupStatus::finished_threads_counters_memory

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
>Lets not rename any methods, only fix leaks and backport it

Great.

Rebased (reorganize patches to separate them more logically) and updated the pull request description (remove some details that regular user do not need, and left this details in the commit descriptions for careful reader).

kitaisreal commented:
Test failures are unrelated.

Algunenano commented:
I see that the 22.3 backport PR is still open (https://github.com/ClickHouse/ClickHouse/pull/40901) but while the backport to 22.8, 22.7 and 22.6 is done already. Is there something missing we can help with to get it to that LTS too?

kitaisreal commented:
@Algunenano there is issues with crash in tests that is related to changes. ThreadPoolRemoteFSReader submit crash https://s3.amazonaws.com/clickhouse-test-reports/40901/8b70f9c77412a25bd597a816a420df7025a352d1/stateful_tests__debug__actions_.html.

kitaisreal commented:
@Algunenano it seems that it was typo, lets wait builds and merge.
